core/src/main/scala/kafka/metrics/clientmetrics/ClientMetricsCache.scala

```diff
@@ -0,0 +1,166 @@
+import kafka.metrics.clientmetrics.ClientMetricsConfig.SubscriptionGroup
+import org.apache.kafka.common.Uuid
+import org.apache.kafka.common.errors.InvalidRequestException
+import org.apache.log4j.helpers.LogLog.error
+
+import java.util.Calendar
+import java.util.concurrent.ConcurrentHashMap
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+/**
+ * Allowed operations on the ClientMetricsCache
+ */
+object ClientMetricsCacheOperation extends Enumeration {
+  type ClientMetricsCacheOperation = Value
+  val CM_SUBSCRIPTION_ADDED, CM_SUBSCRIPTION_DELETED, CM_SUBSCRIPTION_UPDATED, CM_SUBSCRIPTION_TTL = Value
+}
+
+/**
+ * Client Metrics Cache:
+ *   Cache of the client instance state objects that are created in response to the client's
+ *   GetTelemetrySubscriptionRequest message.
+ *
+ *   Eviction Policy:
+ *      Elements stayed too long (beyond the TTL time period) would be cleaned up from the cache by running GC
+ *      which is is an asynchronous task triggered by ClientMetricsManager.
+ *
+ *   Invalidation (keeping the client instance state in sync with client metric subscriptions):
+ *      Whenever a client metric subscription is added/deleted/modified cache is invalidated to make sure that
+ *      cached elements reflects the changes made to the client metric subscriptions.
+ */
+object  ClientMetricsCache {
+  val DEFAULT_TTL_MS = 60 * 1000  // One minute
+  val CM_CACHE_GC_INTERVAL = 5 * 60 * 1000 // 5 minutes
+  val gcTs = Calendar.getInstance.getTime
+
+  private val cacheInstance = new ClientMetricsCache
+  def getInstance = cacheInstance
+  val getSize = getInstance.getSize
+
+  /**
+   * Launches the asynchronous task to clean the client metric subscriptions that are expired in the cache.
+   * TODO: in future, if needed, we may need to run this task periodically regardless of the size of the cache
+   */
+  def runGC()=  {
+    gcTs.synchronized {
+      val currentTime = Calendar.getInstance.getTime
+      if (currentTime.getTime - gcTs.getTime > CM_CACHE_GC_INTERVAL) {
+        gcTs.setTime(currentTime.getTime)
+        cleanupExpiredEntries("GC").onComplete {
+          case Success(value) => info(s"Client Metrics subscriptions cache cleaned up $value entries")
+          case Failure(exception) =>
+            error(s"Client Metrics subscription cache cleanup failed ${exception.getMessage}")
+        }
+      }
+    }
+  }
+
+  private def isExpired(element: CmClientInstanceState) = {
+    val currentTs = Calendar.getInstance.getTime
+    val delta = currentTs.getTime - element.getLastMetricsReceivedTs.getTime
+    delta > Math.max(3 * element.getPushIntervalMs, DEFAULT_TTL_MS)
+  }
+
+  private def cleanupExpiredEntries(reason: String): Future[Int] = Future {
+    val preCleanupSize = cacheInstance.getSize
+    cacheInstance.invalidate(null, null, CM_SUBSCRIPTION_TTL)
```

Comment:
Since I am using concurrent hash map I thought we do not need any extra locking as we only either add/delete elements to the list, however still needs to make sure that what happens if an element is deleted in one thread while another thread is scanning the list.

Reply:
I will do more investigation on the locking requirement.